lib/web/terminal.go

```diff
@@ -284,12 +292,44 @@ func (t *TerminalHandler) makeClient(ws *websocket.Conn) (*client.TeleportClient
 	tc.OnShellCreated = func(s *ssh.Session, c *ssh.Client, _ io.ReadWriteCloser) (bool, error) {
 		t.sshSession = s
 		t.windowChange(&t.params.Term)
+
 		return false, nil
 	}
 
 	return tc, nil
 }
 
+// startPingLoop starts a loop that will continuously send a ping frame through the websocket
+// to prevent the connection between web client and teleport proxy from becoming idle.
+// Interval is determined by the keep_alive_interval config set by user (or default).
+// Loop will terminate when there is an error sending ping frame or when terminal session is closed.
+func (t *TerminalHandler) startPingLoop(ws *websocket.Conn) {
+	// Define our own marshal func to just return a ping payload type.
+	codec := websocket.Codec{Marshal: func(v interface{}) (data []byte, payloadType byte, err error) {
+		return nil, websocket.PingFrame, nil
+	}}
+
+	t.log.Debugf("Starting websocket ping loop with interval %v.", t.params.KeepAliveInterval)
+	tickerCh := time.NewTicker(t.params.KeepAliveInterval)
+	defer tickerCh.Stop()
+
+	for {
+		select {
+		case <-tickerCh.C:
+			// Pongs are internally handled by the websocket library.
+			// https://github.com/golang/net/blob/master/websocket/hybi.go#L291
+			if err := codec.Send(ws, nil); err != nil {
+				t.log.Errorf("Unable to send ping frame to web client: %v.", err)
+				t.terminalCancel()
```

Comment:
should it be `t.Close()`?

Reply:
revised: https://github.com/gravitational/teleport/pull/3675/commits/8eaa9daab376906be142cea8023340e99a6820c2